In [10]:
import os

import numpy as np
import pandas as pd
import pathlib

import ambulance_game as abg

## Number of experiments
- Total in this machine: 5,128,336
- New data in this machine: 26532
- In siren so far: 26570

# Looking around the parameters

In [2]:
df = pd.read_csv(pathlib.Path("../_old_parameters/main.csv"))

In [3]:
len(df)

5078880

In [4]:
df_new = pd.read_csv(pathlib.Path("../data/_parameters/main.csv"))

In [5]:
len(df_new)

51528

# Summary

In [9]:
summary = df.describe()

In [10]:
summary

alpha  buffer_capacity_1  buffer_capacity_2    lambda_1_1  \
count  5.078880e+06       5.078880e+06       5.078880e+06  5.078880e+06   
mean   4.999894e-01       1.992345e+00       2.045931e+00  9.939033e-01   
std    3.027642e-01       8.715865e-02       5.229519e-01  6.087603e-01   
min    0.000000e+00       1.000000e+00       2.000000e+00  0.000000e+00   
25%    2.500000e-01       2.000000e+00       2.000000e+00  4.210526e-01   
50%    5.000000e-01       2.000000e+00       2.000000e+00  9.473684e-01   
75%    7.500000e-01       2.000000e+00       2.000000e+00  1.473684e+00   
max    1.000000e+00       2.000000e+00       8.000000e+00  2.000000e+00   

         lambda_1_2      lambda_2          mu_1          mu_2  \
count  5.078880e+06  5.078880e+06  5.078880e+06  5.078880e+06   
mean   1.115661e+00  4.019762e+00  1.987909e+00  2.021893e+00   
std    1.699545e+00  2.373917e+00  1.376609e-01  2.492627e-01   
min    0.000000e+00  1.000000e-01  4.205707e-01  2.000000e+00   
25%    5.263158e-01  2.006897e+00  2.000000e+00  2.000000e+00   
50%    1.052632e+00  3.913793e+00  2.000000e+00  2.000000e+00   
75%    1.578947e+00  6.093103e+00  2.000000e+00  2.000000e+00   
max    3.401911e+01  8.000000e+00  2.000000e+00  4.859873e+00   

       num_of_servers_1  num_of_servers_2  system_capacity_1  \
count      5.078880e+06      5.078880e+06       5.078880e+06   
mean       1.030621e+00      1.045931e+00       2.045931e+00   
std        3.486346e-01      5.229519e-01       5.229519e-01   
min        1.000000e+00      1.000000e+00       2.000000e+00   
25%        1.000000e+00      1.000000e+00       2.000000e+00   
50%        1.000000e+00      1.000000e+00       2.000000e+00   
75%        1.000000e+00      1.000000e+00       2.000000e+00   
max        5.000000e+00      7.000000e+00       8.000000e+00   

       system_capacity_2        target  
count       5.078880e+06  5.078880e+06  
mean        2.038276e+00  5.000000e+00  
std         4.357932e-01  3.034885e+00  
min         2.000000e+00  0.000000e+00  
25%         2.000000e+00  2.500000e+00  
50%         2.000000e+00  5.000000e+00  
75%         2.000000e+00  7.500000e+00  
max         7.000000e+00  1.000000e+01

In [7]:
summary_new = df_new.describe()

In [8]:
summary_new

alpha  buffer_capacity_1  buffer_capacity_2    lambda_1_1  \
count  26532.000000            26532.0            26532.0  26532.000000   
mean       0.764490                5.0                4.0      2.914179   
std        0.334019                0.0                0.0      2.105936   
min        0.000000                5.0                4.0      0.000000   
25%        0.600000                5.0                4.0      1.500000   
50%        0.950000                5.0                4.0      3.000000   
75%        0.980000                5.0                4.0      4.500000   
max        1.000000                5.0                4.0      6.000000   

         lambda_1_2      lambda_2     mu_1     mu_2  num_of_servers_1  \
count  26532.000000  26532.000000  26532.0  26532.0           26532.0   
mean       2.986318     11.614013      2.0      3.0               3.0   
std        2.125270      7.409679      0.0      0.0               0.0   
min        0.000000      0.100000      2.0      3.0               3.0   
25%        1.500000      5.411111      2.0      3.0               3.0   
50%        3.000000     10.722222      2.0      3.0               3.0   
75%        4.500000     18.688889      2.0      3.0               3.0   
max        6.000000     24.000000      2.0      3.0               3.0   

       num_of_servers_2  system_capacity_1  system_capacity_2        target  
count           26532.0            26532.0            26532.0  26532.000000  
mean                2.0                6.0                7.0      5.000000  
std                 0.0                0.0                0.0      3.162337  
min                 2.0                6.0                7.0      0.000000  
25%                 2.0                6.0                7.0      2.000000  
50%                 2.0                6.0                7.0      5.000000  
75%                 2.0                6.0                7.0      8.000000  
max                 2.0                6.0                7.0     10.000000

# Separation of set of parameters

In [11]:
df["buffer_capacity_2"][5039999]

2

In [12]:
df["buffer_capacity_2"][5040000]

8

# Get hash of given set of parameters

In [33]:
def get_index_of_values(problem_parameters, data, atol=1e-08, rtol=1e-05):
    """
    Get the indices of the rows that match the given parameters' values
    """
    index = data.index
    condition = (
        np.isclose(data["lambda_2"], problem_parameters["lambda_2"], atol=atol, rtol=rtol)
        & np.isclose(data["lambda_1_1"], problem_parameters["lambda_1_1"], atol=atol, rtol=rtol)
        & np.isclose(data["lambda_1_2"], problem_parameters["lambda_1_2"], atol=atol, rtol=rtol)
        & np.isclose(data["mu_1"], problem_parameters["mu_1"], atol=atol, rtol=rtol)
        & np.isclose(data["mu_2"], problem_parameters["mu_2"], atol=atol, rtol=rtol)
        & np.isclose(data["num_of_servers_1"], problem_parameters["num_of_servers_1"])
        & np.isclose(data["num_of_servers_2"], problem_parameters["num_of_servers_2"])
        & np.isclose(data["system_capacity_1"], problem_parameters["system_capacity_1"])
        & np.isclose(data["system_capacity_2"], problem_parameters["system_capacity_2"])
        & np.isclose(data["buffer_capacity_1"], problem_parameters["buffer_capacity_1"])
        & np.isclose(data["buffer_capacity_2"], problem_parameters["buffer_capacity_2"])
        & np.isclose(data["alpha"], problem_parameters["alpha"])
        & np.isclose(data["target"], problem_parameters["target"], atol=atol, rtol=rtol)
    )
    indices = index[condition]
    return indices

In [34]:
def create_directory_with_notebook(data, index_value):
    """
    Create the directory with main.ipynb that contains a brief analysis for that scenario
    """
    hash_value = data[index_value:index_value + 1]["hash_value"]
    dirname = hash_value.to_string(index=False)[1:]
    new_dir = pathlib.Path() / dirname
    new_dir.mkdir(exist_ok=True)
    command = 'copy _main\main.ipynb ' + dirname + '\\'
    os.system(command)
    return dirname

# Experiments with small games

In [ ]:
data = df

# 1) Changing the target

### 1.1) R = 5.2

In [15]:
problem_parameters = {
        "alpha": 0.95,
        "buffer_capacity_1": 2,
        "buffer_capacity_2": 2,
        "lambda_1_1": 1.8,
        "lambda_1_2": 1.2,
        "lambda_2": 4.2,
        "mu_1": 2,
        "mu_2": 2,
        "num_of_servers_1": 1,
        "num_of_servers_2": 1,
        "system_capacity_1": 2,
        "system_capacity_2": 2,
        "target": 5.25,
    }
index = get_index_of_values(
    problem_parameters=problem_parameters, data=data, atol=1e-01
)
data.iloc[list(index)]

alpha  buffer_capacity_1  buffer_capacity_2  lambda_1_1  lambda_1_2  \
2667810   0.95                  2                  2    1.789474    1.157895   
2668230   0.95                  2                  2    1.789474    1.263158   
2676210   0.95                  2                  2    1.894737    1.157895   
2676630   0.95                  2                  2    1.894737    1.263158   

         lambda_2  mu_1  mu_2  num_of_servers_1  num_of_servers_2  \
2667810  4.186207   2.0   2.0                 1                 1   
2668230  4.186207   2.0   2.0                 1                 1   
2676210  4.186207   2.0   2.0                 1                 1   
2676630  4.186207   2.0   2.0                 1                 1   

         system_capacity_1  system_capacity_2    target  \
2667810                  2                  2  5.263158   
2668230                  2                  2  5.263158   
2676210                  2                  2  5.263158   
2676630                  2                  2  5.263158   

                               hash_value  
2667810  000000a094aa728bb6d364142375d4b4  
2668230  bf94ff36c0fa6d3b9ee82460ddd3c797  
2676210  2016f1dfab8b7adfb2ab939d0687e272  
2676630  36ba11dbfa17feb5cfe3fe6e703b1999

In [12]:
create_directory_with_notebook(data=data, index_value=index[0])

'000000a094aa728bb6d364142375d4b4'

### 1.2) R = 0.5

In [39]:
problem_parameters = {
        "alpha": 0.95,
        "buffer_capacity_1": 2,
        "buffer_capacity_2": 2,
        "lambda_1_1": 1.8,
        "lambda_1_2": 1.2,
        "lambda_2": 4.2,
        "mu_1": 2,
        "mu_2": 2,
        "num_of_servers_1": 1,
        "num_of_servers_2": 1,
        "system_capacity_1": 2,
        "system_capacity_2": 2,
        "target": 0.5,
    }
index = get_index_of_values(
    problem_parameters=problem_parameters, data=data, atol=1e-01
)
data.iloc[list(index)]

alpha  buffer_capacity_1  buffer_capacity_2  lambda_1_1  lambda_1_2  \
2667801   0.95                  2                  2    1.789474    1.157895   
2668221   0.95                  2                  2    1.789474    1.263158   
2676201   0.95                  2                  2    1.894737    1.157895   
2676621   0.95                  2                  2    1.894737    1.263158   

         lambda_2  mu_1  mu_2  num_of_servers_1  num_of_servers_2  \
2667801  4.186207   2.0   2.0                 1                 1   
2668221  4.186207   2.0   2.0                 1                 1   
2676201  4.186207   2.0   2.0                 1                 1   
2676621  4.186207   2.0   2.0                 1                 1   

         system_capacity_1  system_capacity_2    target  \
2667801                  2                  2  0.526316   
2668221                  2                  2  0.526316   
2676201                  2                  2  0.526316   
2676621                  2                  2  0.526316   

                               hash_value  
2667801  1fde3123427a69cdf0d62fdf3cdf1e57  
2668221  bf95cdd339f749235e8d4a0e8200e448  
2676201  b193a7f260f370fc87ecb4acb7c902d0  
2676621  e6f1ec3374cb331e9e546b12c823401a

In [14]:
create_directory_with_notebook(data=data, index_value=index[0])

'1fde3123427a69cdf0d62fdf3cdf1e57'

### Comparisons
By comparing the jupyter notebooks in directories 000000a094aa728bb6d364142375d4b4/ and 1fde3123427a69cdf0d62fdf3cdf1e57/ we can observe that by decreasing the value of the target ($R$) there is some changes in the price of anarchy ($PoA$) for different performance measures:

- Waiting Time: $\qquad PoA(R \approx 5) = 2.0 \qquad \qquad PoA(R \approx 0.5) = 1.0$ 
- Blocking Time: $\qquad PoA(R \approx 5) = 1.0 \qquad  \qquad PoA(R \approx 0.5) = 2.5$
- Proportion of individuals lost: $\qquad PoA(R \approx 5) = 1.0 \qquad  \qquad PoA(R \approx 0.5) = 1.9$